gobblin-core/src/main/java/gobblin/source/extractor/hadoop/HadoopFileInputSource.java

```diff
@@ -0,0 +1,258 @@
+   * @param configuration a Hadoop {@link Configuration} object carrying Hadoop configurations
+   * @return a {@link FileInputFormat} instance
+   */
+  @SuppressWarnings("unchecked")
+  protected FileInputFormat<K, V> getFileInputFormat(State state, Configuration configuration) {
+    Preconditions.checkArgument(state.contains(FILE_INPUT_FORMAT_CLASS_KEY));
+    try {
+      return (FileInputFormat<K, V>) ReflectionUtils.newInstance(
+          Class.forName(state.getProp(FILE_INPUT_FORMAT_CLASS_KEY)), configuration);
+    } catch (ClassNotFoundException cnfe) {
+      throw new RuntimeException(cnfe);
+    }
+  }
+
+  /**
+   * Get a {@link HadoopFileInputExtractor} instance.
+   *
+   * @param workUnitState a {@link WorkUnitState} object carrying Gobblin configuration properties
+   * @param recordReader a Hadoop {@link RecordReader} object used to read input records
+   * @param fileSplit the {@link FileSplit} to read input records from
+   * @param readKeys whether the {@link OldApiHadoopFileInputExtractor} should read keys of type {@link #<K>};
+   *                 by default values of type {@link #>V>} are read.
+   * @return a {@link HadoopFileInputExtractor} instance
+   */
+  protected abstract HadoopFileInputExtractor<S, D, K, V> getExtractor(WorkUnitState workUnitState,
+      RecordReader<K, V> recordReader, FileSplit fileSplit, boolean readKeys);
+
+  private String serializeFileSplit(FileSplit fileSplit) throws IOException {
+    Closer closer = Closer.create();
+    try {
+      ByteArrayOutputStream byteArrayOutputStream = closer.register(new ByteArrayOutputStream());
+      DataOutputStream dataOutputStream = closer.register(new DataOutputStream(byteArrayOutputStream));
+      fileSplit.write(dataOutputStream);
+      return BaseEncoding.base64().encode(byteArrayOutputStream.toByteArray());
+    } catch (Throwable t) {
+      throw closer.rethrow(t);
+    } finally {
+      closer.close();
+    }
+  }
+
+  private FileSplit deserializeFileSplit(String fileSplitBytesStr) throws IOException  {
+    Closer closer = Closer.create();
+    try {
+      byte[] fileSplitBytes = BaseEncoding.base64().decode(fileSplitBytesStr);
+      ByteArrayInputStream byteArrayInputStream = closer.register(new ByteArrayInputStream(fileSplitBytes));
+      DataInputStream dataInputStream = closer.register(new DataInputStream(byteArrayInputStream));
+      FileSplit fileSplit = ReflectionUtils.newInstance(FileSplit.class, new Configuration());
+      fileSplit.readFields(dataInputStream);
+      return fileSplit;
+    } catch (Throwable t) {
+      throw closer.rethrow(t);
+    } finally {
+      closer.close();
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private TaskAttemptContext getTaskAttemptContext(Configuration configuration, TaskAttemptID taskAttemptID) {
+    Class<?> taskAttemptContextClass;
+
+    // This is a method learned from Spark (See "org.apache.spark.mapreduce.SparkHadoopMapReduceUtil").
```

Comment:
Yes, this is necessary. The same new Hadoop API is available in Hadoop 1 and 2, but the implementation is different between Hadoop 1 and 2.